l1-contracts/contracts/state-transition/MultisigCommitter.sol

```diff
@@ -0,0 +1,188 @@
+	using EnumerableSetUpgradeable for EnumerableSetUpgradeable.AddressSet;
+
+	/// @dev EIP-712 TypeHash for commitBatchesMultisig
+    bytes32 internal constant COMMIT_BATCHES_MULTISIG_TYPEHASH =
+        keccak256("CommitBatchesMultisig(address chainAddress,uint256 processBatchFrom,uint256 processBatchTo,bytes batchData)");
+
+	// per chain verifier role, only applies if useCustomVerifiers is set
+	bytes32 public constant override COMMIT_VERIFIER_ROLE = keccak256("COMMIT_VERIFIER_ROLE");
+
+	EnumerableSetUpgradeable.AddressSet private sharedVerifiers;
+	uint64 public override sharedSigningThreshold;
+
+	struct ChainConfig {
+		bool useCustomVerifiers; // if we should use per chain COMMIT_VERIFIER_ROLE holders instead of the shared verifier set
+		uint64 signingThreshold; // only applies if useCustomVerifiers is true
+	}
+
+	mapping(address chainAddress => ChainConfig) public chainConfig;
+
+	constructor(address _bridgehubAddr) ValidatorTimelock(_bridgehubAddr) {
+		_disableInitializers();
+	}
+
+	/// @inheritdoc IValidatorTimelock
+	function initialize(address _initialOwner, uint32 _initialExecutionDelay) external override(ValidatorTimelock, IValidatorTimelock) initializer {
+		_validatorTimelockInit(_initialOwner, _initialExecutionDelay);
+		__EIP712_init("MultisigCommitter", "1");
+	}
+
+	/// @inheritdoc IValidatorTimelock
+    function commitBatchesSharedBridge(
+		address _chainAddress,
+        uint256 _processBatchFrom,
+        uint256 _processBatchTo,
+        bytes calldata _batchData
+	) public override(ValidatorTimelock, IValidatorTimelock) {
+		if (getSigningThreshold(_chainAddress) != 0) 
+			revert ChainRequiresVerifiersSignaturesForCommit();
+		super.commitBatchesSharedBridge(_chainAddress, _processBatchFrom, _processBatchTo, _batchData);
+	}
+
+	/// @inheritdoc IMultisigCommitter
+	function commitBatchesMultisig(
+        address chainAddress,
+        uint256 processBatchFrom,
+        uint256 processBatchTo,
+        bytes calldata batchData,
+		address[] calldata signers,
+		bytes[] calldata signatures
+    ) external onlyRole(chainAddress, COMMITTER_ROLE) {
+        bytes32 digest = _hashTypedDataV4(keccak256(abi.encode(COMMIT_BATCHES_MULTISIG_TYPEHASH, chainAddress, processBatchFrom, processBatchTo, keccak256(batchData))));
+		_checkSignatures(chainAddress, signers, signatures, digest);
+		// signatures validated, follow normal commitBatchesSharedBridge flow
+		_recordBatchCommitment(chainAddress, processBatchFrom, processBatchTo);
+		// we cannot use _propagateToZKChain here, because function signature is altered
+		IExecutor(chainAddress).commitBatchesSharedBridge(chainAddress, processBatchFrom, processBatchTo, batchData);
+    }
+
+	/// @inheritdoc IMultisigCommitter
+	function getSigningThreshold(address chainAddress) public override view returns (uint64) {
+		if(chainConfig[chainAddress].useCustomVerifiers)
+			return chainConfig[chainAddress].signingThreshold;
```

Comment:
Nit(gas): It is cheaper to store `chainConfig[chainAddress]` in memory and then use it for checking `useCustomVerifiers` and in returning `signingThreshold`. This way you do one storage read instead of two.